### ISSUE TITLE
International characters: folding behavior?

### ISSUE THREAD

**Issue by danka74** [IHTSDO/snowstorm#65]
Have snowstorm develop branch on commit 8908d8fb8f4a4f4ca58499a9d446e07cd225f3c0 (Aug 15) with a fresh import of SNOMED July 2019 and the Swedish extension of May 2019.

When searching for the string "ångström" using STANDARD searchMode on the findBrowserDescription end point "angstrom" and "ångström" are matched, i.e. characters are folded (as they shouldn't). When using REGEX only "ångström" is matched.

When searching for the string "Lasègues test" using STANDARD searchMode on the findBrowserDescription end point "Lasegues test" and "Lasègues test" are matched, and using REGEX only matches "Lasègues test". Searching for "Lasegues test" (without the accent) using STANDARD matches "Lasègues test" and "Lasegues test" (as it should), and using REGEX does not match anything.

Accept_language=sv for all searches. Using the Swagger interface. This is the CodeSystem:
{
  "shortName": "SNOMEDCT-SE",
  "countryCode": "se",
  "defaultLanguageCode": "sv",
  "branchPath": "MAIN/SNOMEDCT-SE",
  "languages": {
    "sv": "Swedish",
    "en": "English"
  },
...

E.g. http://localhost:8080/browser/MAIN%2FSNOMEDCT-SE/descriptions?term=%C3%A5ngstr%C3%B6m&groupByConcept=false&searchMode=STANDARD&offset=0&limit=50

From my tests it seems that folding is still done for Swedish characters in STANDARD mode and that no folding is done in REGEX mode. I do though see that e.g. the DescriptionServiceTest runs fine...

**Comment by kaicode** [comment]
Hi Daniel,

This is the current behaviour in Snowstorm v4.1.0:

When you request with "Accept Language: sv" Snowstorm adds "en" the default language if it's not already in the list. This is to prevent strange behaviour when content is matched but not translated. So your search is equivalent to "Accept Language: sv,en".

The search criteria will then match either:
- Any Swedish description (language code sv) where the term matches the search term after the Swedish character folding rules are applied. In your example this will only match descriptions with a term containing "ångström*" where * is a wildcard so something like "ångströms" would also match. (This search behaviour allows multiple prefix typeahead.)
- OR any English description where the term matches the search term after English character folding rules are applied (characters folded to "angstrom").

The results are sorted by the shortest term first as normal. This could be improved by scoring Swedish terms higher in this case.

The results include the languageCode of the match along with the FSN and PT of the concept which will also be translated if a translation is available.

It is currently not possible to filter the matches by language. This improvement will be added very soon because it's a glaring omission from the international search. For a workaround please consider filtering the matched descriptions by the Swedish module.

So in summary: the Accept Language header is being used as a hint to control matching and response translation but always includes english. A language parameter will be added in the next few days to filter results completely. I will also add notes for these parameters in the swagger docs because it's not going to be obvious.

Kind regards,
Kai

**Comment by kaicode** [comment]
@danka74 I've now added the language parameter and updated the swagger docs with notes in the develop branch. Please let me know if this is working for you or if any further changes or clarification is needed. As always, thanks for your feedback!

**Comment by kaicode** [comment]
Hi @danka74,

I confirm that the regex search runs against a non-folded version of the description term. 

This is because in Snowstorm we are storing the non-folded term in Elasticsearch as a _keyword_ field type which preserves the whole term. The folded version of the term is being stored in a _text_ field type which breaks up the term and stores each word separately which really boosts performance.

The _keyword_ field type is the only one which can be used for regular expressions which match against multiple words within the term. It would be possible to store a folded version of the term in a third field of type _keyword_, however I am hesitant to do that because it would mean storing every term in Snowstorm three times which would have an impact on index sizes.

Kind regards,
Kai
